[core][cgraph][docs] Compiled Graph troubleshooting (#51030)
ray-project/ray: doc/source/ray-core/compiled-graph/troubleshooting.rst

It's recommended to explicitly delete the NumPy array as Python may not always garbage collect the NumPy array immediately as you may expect.

For example, the following code sample could result in a hang or RayChannelTimeoutError if the NumPy array isn't deleted:

.. literalinclude:: ../doc_code/cgraph_troubleshooting.py
    :language: python
    :start-after: __numpy_troubleshooting_start__
    :end-before: __numpy_troubleshooting_end__

In the preceding code snippet, Python may not garbage collect the NumPy array in `result` on each iteration of the loop. 
Therefore, you should explicitly delete the NumPy array before you try to get the result of subsequent Compiled Graph executions.

Explicitly teardown before reusing the same actors
--------------------------------------------------
If you want to reuse the actors of a Compiled Graph, it's important to explicitly teardown the Compiled Graph before reusing the actors. 
Without explicitly tearing down the Compiled Graph, the resources created for actors in a Compiled Graph may have conflicts with further usage of those actors.

> For example, in the following code, Python could delay garbage collection which triggers the implicit teardown of the first Compiled Graph. This could lead to a segfault due to the resource conflicts mentioned:

Reviewer: ```suggestion
For example, in the following code, Python could delay garbage collection, which triggers the implicit teardown of the first Compiled Graph. This could lead to a segfault due to the resource conflicts mentioned:
```

Troubleshooting

> ==========

Reviewer 2: length not match

This page contains common issues and solutions for Compiled Graph execution.

> Multiple executions with NumPy arrays

Reviewer 2: Simply "Return NumPy arrays"? Compiled Graph execute() is supposed to be called multiple times

> ----------------

Reviewer 2: length not matching

> Ray chooses to zero-copy deserialize NumPy arrays when possible. For multiple Compiled Graph executions, 

Reviewer 2: ```suggestion
Ray zero-copy deserializes NumPy arrays when possible. If a Compiled Graph executes multiple times, 
```

> Ray zero-copy deserializes NumPy arrays when possible. If you execute compiled graph with a NumPy array output multiple times, 

Reviewer 2: QQ: do you know what's the case that Ray does not deserialize NumPy?

Author: I think this https://github.com/ray-project/ray/pull/50731 describes the case

this can lead to issues if a NumPy array from a previous Compiled Graph execution isn't deleted before attempting to get the result 
of a following execution of the same Compiled Graph. This is because the NumPy array stays in the buffer of the Compiled Graph until you or Python delete it. 
It's recommended to explicitly delete the NumPy array as Python may not always garbage collect the NumPy array immediately as you may expect.

> For example, the following code sample could result in a RayChannelTimeoutError:

Reviewer 2: probably better to rephrase, `RayChannelTimeoutError` is used to catch issues like this. It would cause hanging.

.. literalinclude:: ../doc_code/cgraph_troubleshooting.py
    :language: python
    :start-after: __numpy_troubleshooting_start__
    :end-before: __numpy_troubleshooting_end__

In the preceding code snippet, even though you may expect `result` to be garbage collected 

> on each iteration of the loop, there is no guarantee of that from Python. Therefore, you should explicitly delete or 

Reviewer 2: The users don't *expect* GC to happen. They expect the code to run without an issue. I think we should explain why there is an issue here.

Author: reworded a bit, the issue is explained before so I don't think we need to fully re-explain again here

copy and delete the NumPy array before you try to get the result of subsequent Compiled Graph executions.

Explicitly teardown before reusing the same actors

> ----------------

Reviewer 2: ditto

> Therefore, you should explicitly delete or copy and delete the NumPy array before you try to get the result of subsequent Compiled Graph executions.

Reviewer 2: What does "copy and delete" mean? Feel it is not super clear

Author: was trying to say that you can deep copy the numpy array if you want to keep it around but delete the reference you got from the cgraph

I'm gonna just remove the `copy and delete` part, I think a user can figure that out and it makes the sentence more confusing than it needs to me

> Explicitly teardown before reusing the same actors

Reviewer 2: ```suggestion
Explicitly teardown before creating another Compiled Graph
```

Author: it's before reusing the same actors in cgraph or core though right? like we should recommend not using actors in ray core when they're being used in a cgraph

Reviewer 2: Discussed offline, SG

> When reusing the same actors, it's important to explicitly teardown the Compiled Graph before reusing them.

Reviewer 2: nit: tear down?

Author: I think it's more likely for people to search as one word since the api is teardown as one word. But ya looked it up `teardown` is a noun `tear down` for usage as a verb, and it's a verb here so technically tear down is correct but like I think it's fine, the word should follow the API, will defer decision to the docs reviewer.

Reviewer 2: I'm fine with `teardown`.

> If you want to reuse actors after using them in a Compiled Graph, it's important to explicitly teardown the Compiled Graph before reusing the actors. 

Reviewer 2: ```suggestion
If you want to create another Compiled Graph and reuse actors of an existing one, it's important to explicitly teardown the existing Compiled Graph first. 
```

Relying on the previous compiled graph to be garbage collected when expected can lead to issues

> as Python may not always delete a Compiled Graph immediately as you may expect. 

Reviewer 2: The user does not *rely on* the compiled graph to be GC-ed, not needing to call `teardown()` in the general case is our API, and they didn't need to understand what was going on underneath. Can we rephrase here just to explain why they need to call `teardown()`?

Author: so trying to explain why they need to teardown involves explaining that we open channels on actors and that interferes with standard actor behavior or with creating another compiled graph with the same actors. I'm not sure if that's a detail we want to hide from users?

> For example, code such as the following could result in a segfault under specific conditions:

Reviewer 2: explain a bit why the segfault: e.g., previous teardown was not complete then xxx happens.
be more specific on what are the "specific conditions": e.g., the teardown is interleaved with the second compiled graph then xxx happens

Author: added a short explanation

> If you don't explicitly teardown the Compiled Graph, the actors may not behave as expected in future use because Ray opens channels on 

Reviewer 2: Feel free to rephrase, but users don't have the prior knowledge of "channels" so we need to explain in a way they could understand.

> If you don't explicitly teardown the Compiled Graph, the actors may not behave as expected in future use because Ray opens channels on 

Reviewer 2: ```suggestion
Without this, the new Compiled Graph may fail when the old Compiled Graph is being torn down at the same time.
```

Author: it's not necessarily being torn down at the same time, issue is just that we're trying to create a new cgraph on actors where there's already one there

Reviewer 2: SG

> For example, the following code could result in a segfault because Python could delay teardown until after the attempt to reuse the actor for a second Compiled Graph:

Reviewer 2: ```suggestion
For example, the following code could result in a segfault because Python could delay garbage collection (which triggers implicit teardown of the first Compiled Graph) until the point the second Compiled Graph is being created:
```

> For example, the following code could result in a segfault because Python could delay garbage collection (which triggers implicit teardown of the first Compiled Graph) until the point the second Compiled Graph is being created:

Reviewer 2: Feel free to rephrase.
